glusterd2/commands/volumes/brick-replace.go

```diff
@@ -0,0 +1,205 @@
+
+	subvolumes := make([]api.SubvolReq, 0)
+	volreq := api.VolCreateReq{
+		Subvols:      subvolumes,
+		Size:         vol.Capacity,
+		LimitPeers:   req.LimitPeers,
+		LimitZones:   req.LimitZones,
+		ExcludePeers: req.ExcludePeers,
+		ExcludeZones: req.ExcludeZones,
+	}
+	availableVgs, err := bricksplanner.GetAvailableVgs(&volreq)
+	if err != nil {
+		return
+	}
+	// TODO: check for available vgs in zones already being used in volume.
+	if len(availableVgs) == 0 {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "No volume groups are available")
+		return
+	}
+
+	mtabEntries, err := volume.GetMounts()
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get source brick information like size etc
+	brickInfo, err := volume.BrickStatus(srcBrickInfo, mtabEntries)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get new brick from the available vgs
+	newBrick := bricksplanner.GetNewBrick(availableVgs, brickInfo, vol, subVolIndex, brickIndex)
+
+	peerID := uuid.Parse(newBrick.PeerID)
+	if peerID == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "peer id of new brick could not be parsed")
+		return
+	}
+	allPeerIDs := vol.Nodes()
+	nodes := []uuid.UUID{peerID}
+	txn, err := transaction.NewTxnWithLocks(ctx, volname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	defer txn.Done()
+
+	txn.Steps = []*transaction.Step{
+		{
+			DoFunc: "brick-replace.PrepareBricks",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc: "brick-replace.ReplaceVolinfo",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		{
+			DoFunc:   "vol-create.InitBricks",
+			UndoFunc: "vol-create.UndoInitBricks",
+			Nodes:    nodes,
+		},
+		{
+			DoFunc: "vol-start.StartBricks",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc:   "vol-create.StoreVolume",
+			UndoFunc: "vol-create.UndoStoreVolume",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+		},
+		{
+			DoFunc: "vol-expand.NotifyClients",
+			Nodes:  allPeerIDs,
+		},
+	}
+
+	if err := txn.Ctx.Set("newBrick", &newBrick); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	if err := txn.Ctx.Set("srcBrickInfo", &srcBrickInfo); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	if err := txn.Ctx.Set("subVolIndex", &subVolIndex); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	if err := txn.Ctx.Set("brickIndex", &brickIndex); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	if err := txn.Ctx.Set("volinfo", &vol); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if err := txn.Do(); err != nil {
+		status, err := restutils.ErrToStatusCode(err)
```

Comment:
Ideally this `restutils.SendHTTPError(ctx, w, status, err)` should log the error message

Reply:
library function should not log the error.

Reply:
if we log at the higger function it will be defficult to track do where the error has occured, we need to log it in `Do()` incase of failure, currently we are following the same logging style for other function